### PR TITLE
[Serializer] Add new normalizer: PropertyInfoNormalizer based on PropertyInfo / PropertyAccess components

### PR DESCRIPTION
**Pull request by joelwurtz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes (not done yet)
| Tests pass?   | no
| Fixed tickets | #30818 
| License       | MIT
| Doc PR        |  -

This add the new Normalizer, as a Draft as we don't have all the building blocks actually, to show how we want to resolve things.

Will update this PR on the course of others advancement for #30818

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    private $denormalizerExtractor;
+
+    private $propertyAccessor;
+
+    private $nameConverter;
+
+    private $objectClassResolver;
+
+    public function __construct($instantiator, ObjectPropertyListExtractor $normalizerExtractor, PropertyListExtractorInterface $denormalizerExtractor, PropertyAccessorInterface $propertyAccessor, AdvancedNameConverterInterface $nameConverter = null, callable $objectClassReolver = null)
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php]: We should allow to pass the standard NameConverterInterface here (and pass the extra argument required for the advanced one).

File: src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        return \class_exists($type) && \is_array($data);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $properties = $this->normalizerExtractor->getProperties($object, $context);
+        $class = $this->objectClassResolver ? ($this->objectClassResolver)($object) : \get_class($object);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php]: Shouldn’t we deal with Doctrine proxies here?

**Review by dunglas — CHANGES_REQUESTED** [review]
Very promising!

File: src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php (hunk 2 above)
Comment (joelwurtz) [reply src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php]: That's what we do no ?

Allowing an object class resolver will allow to inject a specific callback to handle this problem.

**Comment by Nyholm** [comment]
What is the status of this PR?

@joelwurtz could you finish this or leave a comment to help others taking over this PR?

**Comment by joelwurtz** [comment]
This PR need the advancement of all points present in https://github.com/symfony/symfony/issues/30818 as it will use all interface / implementations make on those differents PR to create this new normalizer

File: src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        $data = [];
+
+        foreach ($properties as $property) {
+            $propertyName = $this->nameConverter ? $this->nameConverter->normalize($property, $class, $format, $context) : $property;
+            $value = $this->propertyAccessor->getValue($object, $property);
+
+            if (!is_scalar($value)) {
```
Comment (yceruto) [inline src/Symfony/Component/Serializer/Normalizer/PropertyInfoNormalizer.php:100]: I'd add `null !== $value &&` before to save some iterations, wdyt?

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Pull request closed without merging**